content/article/enhancing-midi-hardware-with-perl.md

```diff
@@ -0,0 +1,510 @@
+
+The FL Studio version also has a downside. The played sequence only shows the
+played notes in the piano roll, not the additional fifth and pedal tone. Tweaking timing and
+velocity, or adding additional melody is not trivial - any additional notes in the piano
+roll will play three notes in the Patcher instrument.
+
+If we could coax our MIDI device into producing these additional notes,
+there would be no need for tricky patching plus we might end up with a more
+flexible result.
+
+### Perl Tone
+
+The approach described here will set up a new software-defined MIDI device
+which will proxy events from our hardware, while applying any number of
+filters to events before they are forwarded. These examples will make use of
+[Perl bindings to RtMidi](https://metacpan.org/pod/MIDI::RtMidi::FFI::Device).
+
+We're going to need a little bit of framework code to get started. While the
+[simplest RtMidi callback examples](https://github.com/jbarrett/MIDI-RtMidi-FFI/blob/main/examples/simple_callback.pl)
+just sleep to let the RtMidi event loop take over, we may wish to schedule our
+own events later. I went into some detail previously on
+[Perl, IO::Async, and the RtMidi event loop](https://fuzzix.org/perl-ioasync-and-the-rtmidi-event-loop).
+
+The framework will need to set up an event loop, manage two or more MIDI
+devices, and store some state to influence decision-making within filter
+callback functions. Let's start with those:
+
+```perl
+class MidiFilter {
+    field $loop       = IO::Async::Loop->new;
+    field $midi_ch    = IO::Async::Channel->new;
+    field $midi_out   = RtMidiOut->new;
+    field $input_name = $ARGV[0];
+    field $filters    = {};
+    field $stash      = {};
+```
+
+Aside from our event `$loop` and `$midi_out` device, there are fields for
+getting `$input_name` from the command line, a `$stash` for communication
+between callbacks and a store for callback `$filters`. The callback store will hold
+callbacks keyed on MIDI event names, e.g. "note\_on". The channel `$midi_ch`
+will be used to receive events from the MIDI input controller.
+
+Methods for creating new filters and accessing the stash are as follows:
+
+```perl
+    method add_filter( $event_type, $action ) {
+        push $filters->{ $event_type }->@*, $action;
+    }
+
+    method stash( $key, $value = undef ) {
+        $stash->{ $key } = $value if defined $value;
+        $stash->{ $key };
+    }
+```
+
+Adding a filter requires an event type, plus a callback. Callbacks are pushed
+into `$filters` for each event type in the order they are declared.
+If a `$value` is supplied while accessing the stash, it will be stored for the
+given `$key`. The value for the given `$key` is returned in any case.
+
+Let's add some methods for sending MIDI events:
+
+```perl
+    method send( $event ) {
+        $midi_out->send_event( $event->@* );
+    }
+
+    method delay_send( $dt, $event ) {
```

Comment:
Nit-pick: I'd use a meaningful word instead of a mystery abbreviation for `$dt`, which often means a DateTime object...